### ISSUE TITLE
Allow different onDelete behavior for inverse side of ManyToMany

### ISSUE THREAD

**Issue by csuich2** [typeorm/typeorm#3505]
**Issue type:**

[ ] question
[ ] bug report
[x] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[ ] `latest`
[x] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

ManyToMany relationships currently use the same onDelete option for both directions of the join table. It would be nice to be able to specify a different onDelete option for the inverse side of a ManyToMany relation.

For example: If I have ```Post``` which have a ManyToMany relationship with ```Category```, I may want the deletion of a ```Post``` to have an no action on delete, while deletion of a ```Category``` would cascade delete all ```Post```s which have that category.

I believe this could be fixed an ```onDeleteInverse``` option to ```JoinTableOptions``` and ```JoinTableMultipleColumnsOptions``` so that ```JunctionEntityMetadataBuilder.build()``` can create the foreign keys with two different ```ON DELETE``` behaviors.

**Comment by pleerock** [comment]
We need a PR for this feature.

**Comment by csuich2** [comment]
I'm happy to open the PR, I just wanted to talk through the solution with you before opening it. Do you think this is the right approach?

**Comment by pleerock** [comment]
currently we have no action onDelete right?

**Comment by csuich2** [comment]
Yeah, JoinColumn has no onDelete property. I believe ```JunctionEntityMetadataBuilder.build)``` uses the onDelete option from one side of the relation metadata. I'm not sure which, but I assume the side with the JoinColumn on it.

**Comment by pleerock** [comment]
no, it shouldn't use any other's onDelete.

**Comment by csuich2** [comment]
I'm on @next, so my line numbers might be a bit different, but this is what I've found so far:

```JunctionEntityMetadataBuilder``` line 130:
``` typescript
        // create junction table foreign keys
        entityMetadata.foreignKeys = [
            new ForeignKeyMetadata({
                entityMetadata: entityMetadata,
                referencedEntityMetadata: relation.entityMetadata,
                columns: junctionColumns,
                referencedColumns: referencedColumns,
                onDelete: relation.onDelete ||  "CASCADE"
            }),
            new ForeignKeyMetadata({
                entityMetadata: entityMetadata,
                referencedEntityMetadata: relation.inverseEntityMetadata,
                columns: inverseJunctionColumns,
                referencedColumns: inverseReferencedColumns,
                onDelete: relation.onDelete || "CASCADE"
            }),
        ];
```
the foreign keys are built here and you can see that both FK's onDelete is set with ```onDelete: relation.onDelete || "CASCADE"```, where relation is the ```RelationMetadata``` from the ```@JoinColumn``` side of the relationship.

I think the right solution is the use the ```onDelete``` property from the inverse side relation, if one is specified, but when debugging there, I've found that ```relation.inverseRelation``` is undefined, even when I know it isn't.

Digging in to that revealed the only place ```relation.inverseRelation``` is set is at ```EntityMetadataBuilder``` line 568:
``` typescript
    /**
     * Computes entity metadata's relations inverse side properties.
     */
    protected computeInverseProperties(entityMetadata: EntityMetadata, entityMetadatas: EntityMetadata[]) {
        entityMetadata.relations.forEach(relation => {

            // compute inverse side (related) entity metadatas for all relation metadatas
            const inverseEntityMetadata = entityMetadatas.find(m => m.target === relation.type || (typeof relation.type === "string" && m.targetName === relation.type));
            if (!inverseEntityMetadata)
                throw new Error("Entity metadata for " + entityMetadata.name + "#" + relation.propertyPath + " was not found. Check if you specified a correct entity object and if it's connected in the connection options.");

            relation.inverseEntityMetadata = inverseEntityMetadata;
            relation.inverseSidePropertyPath = relation.buildInverseSidePropertyPath();

            // and compute inverse relation and mark if it has such
            relation.inverseRelation = inverseEntityMetadata.relations.find(foundRelation => foundRelation.propertyPath === relation.inverseSidePropertyPath);
        });
    }
```

 It will be set if we find a matching ```propertyPath``` which depends on ```RelationMetadata.buildInverseSidePropertyPath()``` building it. However, ```RelationMetadata.buildInverseSidePropertyPath()``` appears to only build the ```inverseRelation``` if ```givenInverseSidePropertyFactory``` is specified by a user:
``` typescript
    /**
     * Builds inverse side property path based on given inverse side property factory.
     * This builder method should be used only after properties map of the inverse entity metadata was build.
     */
    buildInverseSidePropertyPath(): string {

        if (this.givenInverseSidePropertyFactory) {
            const ownerEntityPropertiesMap = this.inverseEntityMetadata.propertiesMap;
            if (typeof this.givenInverseSidePropertyFactory === "function")
                return this.givenInverseSidePropertyFactory(ownerEntityPropertiesMap);

            if (typeof this.givenInverseSidePropertyFactory === "string")
                return this.givenInverseSidePropertyFactory;

        } else if (this.isTreeParent && this.entityMetadata.treeChildrenRelation) {
            return this.entityMetadata.treeChildrenRelation.propertyName;

        } else if (this.isTreeChildren && this.entityMetadata.treeParentRelation) {
            return this.entityMetadata.treeParentRelation.propertyName;
        }

        return "";
    }
```

**Comment by AlexMesser** [comment]
fixed in #5714
